README.md
# tinybird-beam

A Tinybird Apache Beam connector to ingest from an Apache Beam pipeline running in DataFlow, Flink or Spark to a [Tinybird](https://www.tinybird.co/) Data Source

Directories:

- `tinybird`: Contains the source code for the Apache Beam connector
- `dataflow`: Pipeline definition to stream data from Google PubSub to Tinybird using the Apache Beam connector

## Installation

Install the `tinybird-beam` [PyPI package](https://pypi.org/project/tinybird-beam/) in your Python project. See the `dataflow` directory for a usage example.

```sh
pip install tinybird-beam
```

## How to use it

```python
# Import the sink class
from tinybird.beam import WriteToTinybird

# pipe a PCollection with elements of type `Iterable[Dict[str, Any]]` to the sink class
# you should batch your elements beforehand using a `beam.WindowInto` or `beam.transforms.util.GroupIntoBatches`
host = 'https://api.tinybird.co'
token = '' # get it from https://ui.tinybird.co/tokens
datasource = '' # name of the Data Source
columns = '' # comma separated list with the name of the columns of the Data Source
out | "Stream to Tinybird" >> WriteToTinybird(host, token, datasource, columns)
```

## Development

```sh
python3 -m venv env
source env/bin/activate
pip install -e .
```

## DataFlow example

The `dataflow` directory contains some sample code to deploy to DataFlow an Apache Beam pipeline that gets data from a PubSub topic and ingests to a BigQuery table and a Tinybird Data Source.

Follow these steps to run the example:

### Prepare the environment

- Push this Data Source to your Tinybird account and name it `pubsub__invoices`:

```
SCHEMA >
    `id` UInt32,
    `agent_id` UInt8,
    `recipient_code` UInt32,
    `client_id` UInt32,
    `amount` Float32,
    `currency` LowCardinality(String),
    `created_at` DateTime,
    `added_payments` String
```

- Create the Python environment:

```sh
cd dataflow
python3 -m venv env
source env/bin/activate
pip install -r requirements.txt
# Update the variables in the `sample.env` file and source it:
source sample.env
```

- Create the PubSub topic:

```sh
gcloud pubsub topics create demo-topic
```

- Import the `dataflow/pubsub/invoices_sample.json` file to BigQuery. We use as `dataset.table_name` -> `tinybird.pubsub__invoices`.

### Run the example

- Push the Apache Beam pipeline to DataFlow:

```sh
python dataflow.py \
  --project=$PROJECT_NAME \
  --region=$REGION \
  --runner=DataflowRunner \
  --temp_location=$TMP_LOCATION \
  --input_topic=projects/$PROJECT_NAME/topics/$TOPIC \
  --bq_table=tinybird.pubsub__invoices \
  --batch_size=10000 \
  --batch_seconds=30 \
  --batch_key= \
  --tb_host=https://api.tinybird.co \
  --tb_token=$TB_TOKEN \
  --tb_datasource=pubsub__invoices \
  --tb_columns="id,agent_id,recipient_code,client_id,amount,currency,created_at,added_payments"
```

This pipeline will batch 10000 elements or a window of 30 seconds to the Tinybird Data Source.

Once running you'll see this log (it might take several minutes to deploy):

```sh
INFO:apache_beam.runners.dataflow.dataflow_runner:Job 2021-01-22_08_13_37-6419261550197294468 is in state JOB_STATE_RUNNING
INFO:apache_beam.runners.dataflow.dataflow_runner:2021-01-22T16:13:48.981Z: JOB_MESSAGE_DETAILED: Pub/Sub resources set up for topic 'projects/---/topics/demo-topic'.
INFO:apache_beam.runners.dataflow.dataflow_runner:2021-01-22T16:13:50.023Z: JOB_MESSAGE_DEBUG: Starting worker pool setup.
INFO:apache_beam.runners.dataflow.dataflow_runner:2021-01-22T16:13:50.034Z: JOB_MESSAGE_BASIC: Starting 1 workers in europe-west3-b...
INFO:apache_beam.runners.dataflow.dataflow_runner:2021-01-22T16:13:50.048Z: JOB_MESSAGE_DEBUG: Starting worker pool setup.
INFO:apache_beam.runners.dataflow.dataflow_runner:2021-01-22T16:14:17.581Z: JOB_MESSAGE_DETAILED: Autoscaling: Raised the number of workers to 1 so that the pipeline can catch up with its backlog and keep up with its input rate.
INFO:apache_beam.runners.dataflow.dataflow_runner:2021-01-22T16:15:02.264Z: JOB_MESSAGE_DETAILED: Workers have started successfully.
INFO:apache_beam.runners.dataflow.dataflow_runner:2021-01-22T16:15:02.278Z: JOB_MESSAGE_DETAILED: Workers have started successfully.
```

- Start the PubSub publisher:

```
cd pubsub
python pubsub.py
```

After the pipeline is deployed [the job](https://console.cloud.google.com/dataflow/jobs) would look like this:

![](dataflow/dataflow.png)

Once it starts receiving data, you can check the data in your BigQuery table and Tinybird Data Source.

### Clean resources

Stop the DataFlow pipeline from the [Google Cloud console](https://console.cloud.google.com/dataflow/jobs).

Clean up the PubSub topic.

```sh
gcloud pubsub topics delete demo-topic
```


changelog.rst

Tinybird CLI
============

The Tinybird Apache Beam Sink allows you to ingest from an Apache Beam pipeline to a Tinybird Data Source.

Changelog
---------

1.0.0b1
*******

- First version to stream from DataFlow to Tinybird in batches


tinybird/beam.py
import apache_beam as beam
from typing import Any, Iterable, Dict, List
import json

DEFAULT_HOST = 'https://api.tinybird.co'


class InvalidElementType(TypeError):
    """Error Class for when an Element is not the correct type"""


class _Insert(beam.DoFn):
    """An internal DoFn to be used in a PTransform. Not for external use.
    """
    def __init__(self, data_source: str, host: str, token: str, columns: List):
        beam.DoFn.__init__(self)
        self.host = host
        self.token = token
        self.data_source = data_source
        self.columns = columns

    def process(self, element: Iterable[Dict[str, Any]]) -> None:
        try:
            if not isinstance(element, Iterable):
                raise InvalidElementType(f"Wrong element type. Expected Iterable[Dict[str, Any]], received {type(element)}")

            # imports need to be here for serialization purposes
            import requests
            import csv

            from io import StringIO
            from requests.adapters import HTTPAdapter
            from urllib3.util.retry import Retry

            mode = 'append'
            datasource = self.data_source
            token = self.token

            url = f'{self.host}/v0/datasources?mode={mode}&name={datasource}'

            retry = Retry(total=5, backoff_factor=10)
            adapter = HTTPAdapter(max_retries=retry)
            _session = requests.Session()
            _session.mount('http://', adapter)
            _session.mount('https://', adapter)

            csv_chunk = StringIO()
            writer = csv.writer(csv_chunk, delimiter=',', quotechar='"', quoting=csv.QUOTE_NONNUMERIC)

            for row in element:
                rr = []
                for column in self.columns:
                    rr.append(row[column])
                writer.writerow(rr)

            data = csv_chunk.getvalue()
            headers = {
                'Authorization': f'Bearer {token}',
                'X-TB-Client': 'beam-0.1',
            }

            response = _session.post(url, headers=headers, files=dict(csv=data))
            ok = response.status_code < 400
            if not ok:
                raise Exception(json.dumps(response.json()))
        except Exception as e:
            raise e


class WriteToTinybird(beam.PTransform):
    def __init__(self, host: str, token: str, data_source: str, columns: List, **kwargs):
        beam.PTransform.__init__(self)
        self.host = host or DEFAULT_HOST
        self.token = token
        self.data_source = data_source
        self.columns = columns
        self.kwargs = kwargs

    def expand(self, pcoll):
        return (
            pcoll
            | beam.ParDo(_Insert(self.data_source, self.host, self.token, self.columns))
        )


class WriteToTinybirdBatch(beam.PTransform):
    def __init__(self, host: str, token: str, data_source: str, columns: List, batch_size: int = 1000, **kwargs):
        beam.PTransform.__init__(self)
        self.host = host
        self.token = token
        self.data_source = data_source
        self.batch_size = batch_size
        self.columns = columns
        self.kwargs = kwargs

    def expand(self, pcoll):
        return (
            pcoll
            | beam.BatchElements(min_batch_size=self.batch_size, max_batch_size=self.batch_size, **self.kwargs)
            | beam.ParDo(_Insert(self.data_source, self.host, self.token, self.columns))
        )


dataflow/dataflow.py
import argparse
import logging
import json

import apache_beam as beam
from apache_beam.options.pipeline_options import PipelineOptions
from tinybird.beam import WriteToTinybird


class GroupWindowsIntoBatches(beam.PTransform):
    def __init__(self, batch_key, batch_size, max_buffering_duration_secs=60):
        self.batch_key = batch_key
        self.batch_size = int(batch_size)
        self.max_buffering_duration_secs = int(max_buffering_duration_secs)

    def expand(self, pcoll):
        return (
            pcoll | "Generate tuple" >> beam.Map(lambda x: (x.get(self.batch_key, None), x))
                  | "Group into batches"
                    >> beam.transforms.util.GroupIntoBatches(self.batch_size, self.max_buffering_duration_secs)
                  | "Abandon Dummy Key" >> beam.MapTuple(lambda _, val: val)
        )


def run(input_topic=None,
        batch_size=None,
        batch_seconds=None,
        batch_key=None,
        project=None,
        temp_location=None,
        region=None,
        bq_table=None,
        tb_host=None,
        tb_token=None,
        tb_datasource=None,
        tb_columns=None):
    # `save_main_session` is set to true because some DoFn's rely on
    # globally imported modules.
    pipeline_options = PipelineOptions(
        pipeline_args,
        streaming=True,
        save_main_session=True,
        requirements_file='./requirements.txt',
        project=project,
        temp_location=temp_location,
        region=region,
    )

    with beam.Pipeline(options=pipeline_options) as pipeline:
        out = (
            pipeline
            | "Read From PubSub" >> beam.io.ReadFromPubSub(topic=input_topic)
            | "Remove PII from Input" >> beam.Map(lambda payload: json.loads(payload))
        )

        out | "Write to BigQuery" >> beam.io.WriteToBigQuery(
                table=f'{project}:{bq_table}',
                # schema=table_schema,
                write_disposition=beam.io.BigQueryDisposition.WRITE_APPEND,
                # create_disposition=beam.io.BigQueryDisposition.CREATE_IF_NEEDED,
                custom_gcs_temp_location=temp_location
            )

        (out | "Group into batches" >> GroupWindowsIntoBatches(batch_key, batch_size, batch_seconds)
             | "Stream to Tinybird" >> WriteToTinybird(tb_host, tb_token, tb_datasource, tb_columns)
        )


if __name__ == "__main__":  # noqa
    logging.getLogger().setLevel(logging.INFO)

    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--input_topic",
        help="The Cloud Pub/Sub topic to read from.\n"
        '"projects/<PROJECT_NAME>/topics/<TOPIC_NAME>".',
    )
    parser.add_argument(
        "--batch_size",
        type=int,
        default=1000,
        help="Number of elements to batch. If batch_key is provided, batches are per key",
    )
    parser.add_argument(
        "--batch_seconds",
        type=int,
        default=30,
        help="Number of seconds to batch if batch_size is not completed",
    )
    parser.add_argument(
        "--batch_key",
        type=str,
        default=None,
        help="Key of each element to partition the data. Leave to None to not partition",
    )
    parser.add_argument(
        '--temp_location',
        dest='temp_location')
    parser.add_argument(
        '--bq_table',
        dest='bq_table',
        help='Bigquery table path (should already exist). Ex: dataset.table')
    parser.add_argument(
        '--project',
        dest='project',
        help='Google Cloud Platform project ID')
    parser.add_argument(
        '--region',
        dest='region',
        help='GCS region')
    parser.add_argument(
        '--tb_host',
        dest='tb_host',
        help='Tinybird host')
    parser.add_argument(
        '--tb_token',
        dest='tb_token',
        help='Tinybird token')
    parser.add_argument(
        '--tb_datasource',
        dest='tb_datasource',
        help='Tinybird Data Source name')
    parser.add_argument(
        '--tb_columns',
        dest='tb_columns',
        help='Comma separated list of columns in the destination Data Source')
    known_args, pipeline_args = parser.parse_known_args()

    # import ipdb; ipdb.set_trace(context=30)
    run(
        input_topic=known_args.input_topic,
        batch_size=known_args.batch_size,
        batch_seconds=known_args.batch_seconds,
        batch_key=known_args.batch_key,
        project=known_args.project,
        temp_location=known_args.temp_location,
        region=known_args.region,
        bq_table=known_args.bq_table,
        tb_host=known_args.tb_host,
        tb_token=known_args.tb_token,
        tb_datasource=known_args.tb_datasource,
        tb_columns=known_args.tb_columns.split(',')
    )
